pkg/engine/supervisor.go

```diff
@@ -0,0 +1,381 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/logrusorgru/aurora"
+	"github.com/otiai10/copy"
+	"github.com/testground/testground/pkg/api"
+	"github.com/testground/testground/pkg/config"
+	"github.com/testground/testground/pkg/logging"
+	"github.com/testground/testground/pkg/rpc"
+	"github.com/testground/testground/pkg/task"
+	"golang.org/x/sync/errgroup"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type RunInput struct {
+	*api.RunRequest
+	Sources *api.UnpackedSources
+}
+
+type BuildInput struct {
+	*api.BuildRequest
+	Sources *api.UnpackedSources
+}
+
+func (e *Engine) startSupervisor(workers int) {
+	for i := 0; i < workers; i++ {
+		go e.worker(i)
+	}
+}
+
+func (e *Engine) worker(n int) {
+	logging.S().Infow("supervisor worker started", "worker_id", n)
+
+	var (
+		store = e.TaskStorage()
+		queue = e.TaskQueue()
+	)
+
+	for {
+		tsk, err := queue.Pop()
+		if err == task.ErrQueueEmpty {
+			time.Sleep(time.Second)
+			continue
+		}
+
+		ctx, _ := context.WithTimeout(context.Background(), time.Minute*30)
```

Comment:
I think we should `cancel` this context if the task completes successfully, before next iteration.